### PR TITLE
Added recipe for cminpack

### PR DESCRIPTION
**Pull request by rainman110**
Specify library name and version:  **cminpack/1.3.8**

CMinPack is a mathematical solver used in numerical computing. It is quite popular, hence a conan package is yet missed.

The test package uses an official example from the cminpack repository.

The official cmake targets are `cminpack::cminpack` for the double precision library and `cminpack::cminpacks` for the single precision version of it.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
@rainman110 Thank you for your contribution. When doing a draft, please, mark it on Github PR page, so it will be converted automatically:

Screenshot 2022-09-01 at 13-40-13 DRAFT Added recipe for cminpack by rainman110 · Pull Request #12762 · conan-io_conan-center-index

Just adding to the title, does not block the merge button, but once converted to draft state, Github does not allow accidental merge.

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:17]: It does not fit an official SPDX license, but there is no exception. So, BSD-like sounds okay here.

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:35]:
```suggestion
```
Not needed, because this recipe has no other dependencies.

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:1]:
```suggestion
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
```

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:70]: Have you tried as cached_variable on CMakeToolchain instead? variables vs cached_variables are different. Usually, cached_variables should word, we have other cases where we needed to customize a cmake policy using CMakeToolchain.cached_variables and worked.

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:82]:
```suggestion
    def _library_postfix(self):
```
Any custom method should be protected, to avoid any possible name collision with future conan features

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:95]:
```suggestion
        self.cpp_info.libs = ['cminpack' + self._library_postfix()]
```

**Review by uilianries — COMMENTED** [review]
Please, add a test_v1_package folder, to validate Conan 1.x support.

There is a good template here, which you can re-use: https://github.com/conan-io/conan-center-index/tree/master/docs/package/all/test_v1_package

**Comment by rainman110** [comment]
> @rainman110 Thank you for your contribution. When doing a draft, please, mark it on Github PR page, so it will be converted automatically:
> 
> Screenshot 2022-09-01 at 13-40-13 DRAFT Added recipe for cminpack by rainman110 · Pull Request #12762 · conan-io_conan-center-index
> 
> Just adding to the title, does not block the merge button, but once converted to draft state, Github does not allow accidental merge.

Thanks for the hint! I was searching the draft button, but apparently its location has changed since I used this last time.

File: recipes/cminpack/all/conanfile.py
Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:35]: True, will do

File: recipes/cminpack/all/conanfile.py
Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:70]: Not yet... will try this.

File: recipes/cminpack/all/conanfile.py
Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:82]: Good point!

File: recipes/cminpack/all/conanfile.py
Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:70]: This really seems to work. Thanks!

File: recipes/cminpack/all/conanfile.py
Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:16]: Yeah... I basically quoted the text from the homepage 

> Minpack is licensed under a BSD-like license (available in the distribution). .... It [Cminpack]is distributed under the original Minpack license (see the file CopyrightMINPACK.txt in the distribution).

**Comment by rainman110** [comment]
@uilianries I could again need your help here. I figured out, that both executables in the test package link to the target cminpack::cminpacks (which is the single precision library), alltough actually want to link one against the double precision library.

The situation is the following:
 - the package name is "cminpack"
 - I intended to create a target cminpack::cminpack, which is libcminpack.so - the double precision library
 - I also create the target cminpack::cminpacks, which is libcminpacks.so - the single precision library
 - To my surprise, when using `target_link_libraries(... cminpack::cminpack)`, libcminpacks.so is actually linked (instead of libcminpack.so)

Before, I had both libraries as components ( not just cminpacks), but conan did not like  to have a component which overwrites the cminpack::cminpack default target.

How can I fix this?

**Comment by uilianries** [comment]
@rainman110 I will check, but we will need to do a trick with components

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:90]:
```suggestion
        self.cpp_info.components["_cminpack"].libs = ['cminpack' + self._library_postfix()]
        self.cpp_info.components["_cminpack"].includedirs.append(minpack_include_dir)
        self.cpp_info.components["_cminpack"].set_property("cmake_target_name", "cminpack::cminpack")

        self.cpp_info.components["_cminpack"].names["cmake_find_package"] = "cminpack"
        self.cpp_info.components["_cminpack"].names["cmake_find_package_multi"] = "cminpack"

```
Here, you create a new component named `_cminpack`, but you say to use specific names for cmake targets, which is the same from the project name, and it only uses the double precise library

Probably you will need to update system_libs, frameworks and cpp_info.defines for each component, as Conan does not like to mix general cpp_info with components.

Comment (rainman110) [reply recipes/cminpack/all/conanfile.py]: I committed your proposed changes. It fixes linking the correct library in case of the test_v1_package but still links the wrong library in the double precision test case. 

It seems, that the single precision component is still associated to the default cminpack::cminpack target. Is there any way to clear the default target or remove a component from the default target?

Comment (rainman110) [reply recipes/cminpack/all/conanfile.py]: This is the output from the test executables from the v1 package:

```
cminpack/1.3.8@test/testing (test package): Running test()

      fvec

           -1.181606      -1.429655      -1.606344
           -1.745269      -1.840654      -1.921586
           -1.984141      -2.022537      -2.468977
           -2.827562      -3.473582      -4.437612
           -6.047662      -9.267761      -18.91806
      fvecp - fvec

       -7.724666e-09  -3.432406e-09  -2.034843e-10
        2.313685e-09   4.331078e-09   5.984096e-09
        7.363281e-09    8.53147e-09   1.488591e-08
         2.33585e-08   3.522012e-08   5.301255e-08
         8.26666e-08   1.419747e-07    3.19899e-07
      err

           0.1141397     0.09943516     0.09674474
          0.09980447      0.1073116      0.1220445
           0.1526814              1              1
                   1              1              1
                   1              1              1

      fvec

           -1.181606      -1.429655      -1.606344
           -1.745269      -1.840654      -1.921586
           -1.984141      -2.022537      -2.468976
           -2.827562      -3.473582      -4.437612
           -6.047661      -9.267761      -18.91806
      fvecp - fvec

       -0.0001790524  -7.963181e-05  -4.768372e-06
        5.352497e-05   9.989738e-05   0.0001385212
        0.0001702309   0.0001974106   0.0003447533
        0.0005407333   0.0008158684    0.001227856
         0.001914024    0.003288269    0.007408142
      err

           0.2581466      0.2248907              0
            0.225696      0.2429258      0.2760358
           0.3457333              1              1
                   1              1              1
                   1              1              1
``` 

Note the difference in the second column, where the residual is 1e-9 for the double precision case, and ~1e-5 for the single precision case.

If I run the test_package (v2), I get exactly the same output (i.e. the single precision library is linked in both cases):

```
cminpack/1.3.8@test/testing (test package): Running test()
Capturing current environment in C:\src\conan-center-index\recipes\cminpack\all\test_package\build\generators\deactivate_conanrunenv-release-x86_64.bat
Configuring environment variables

      fvec

           -1.181606      -1.429655      -1.606344
           -1.745269      -1.840654      -1.921586
           -1.984141      -2.022537      -2.468976
           -2.827562      -3.473582      -4.437612
           -6.047661      -9.267761      -18.91806
      fvecp - fvec

       -0.0001790524  -7.963181e-05  -4.768372e-06
        5.352497e-05   9.989738e-05   0.0001385212
        0.0001702309   0.0001974106   0.0003447533
        0.0005407333   0.0008158684    0.001227856
         0.001914024    0.003288269    0.007408142
      err

           0.2581466      0.2248907              0
            0.225696      0.2429258      0.2760358
           0.3457333              1              1
                   1              1              1
                   1              1              1
Capturing current environment in C:\src\conan-center-index\recipes\cminpack\all\test_package\build\generators\deactivate_conanrunenv-release-x86_64.bat
Configuring environment variables

      fvec

           -1.181606      -1.429655      -1.606344
           -1.745269      -1.840654      -1.921586
           -1.984141      -2.022537      -2.468976
           -2.827562      -3.473582      -4.437612
           -6.047661      -9.267761      -18.91806
      fvecp - fvec

       -0.0001790524  -7.963181e-05  -4.768372e-06
        5.352497e-05   9.989738e-05   0.0001385212
        0.0001702309   0.0001974106   0.0003447533
        0.0005407333   0.0008158684    0.001227856
         0.001914024    0.003288269    0.007408142
      err

           0.2581466      0.2248907              0
            0.225696      0.2429258      0.2760358
           0.3457333              1              1
                   1              1              1
                   1              1              1
```

File: recipes/cminpack/all/conanfile.py
Comment (prince-chrismc) [reply recipes/cminpack/all/conanfile.py:16]: Please follow https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#what-license-should-i-use-for-a-custom-project-specific-license 🙏

Comment (prince-chrismc) [reply recipes/cminpack/all/conanfile.py:16]:
```suggestion
    license = "LicenseRef-CopyrightMINPACK.txt"
```

File: recipes/cminpack/all/conanfile.py
Comment (prince-chrismc) [inline recipes/cminpack/all/conanfile.py:102]: I thought these were duplicates are first glance... 🙈 

Any objections to being a little more verbose?

```suggestion
            self.cpp_info.components['cminpack_double'].system_libs.append("m")
            self.cpp_info.components['cminpack_single'].system_libs.append("m")
```

File: recipes/cminpack/all/conanfile.py
Comment (prince-chrismc) [inline recipes/cminpack/all/conanfile.py:72]: There should be pkg-config component names declared below as well please

**Review by prince-chrismc — COMMENTED** [review]
Please update the license attribute 🙏 other comments are very minor suggestions but otherwise i'd ✔️

**Comment by rainman110** [comment]
@prince-chrismc thank you for the positive review. I will address your suggestions.

However, there still is the issue linking the wrong library with the new cmaketoolchain generator in the test_package (see comments above), which I am not able to revolve by myself. Do you have any ideas, how to fix this?

**Comment by uilianries** [comment]
@rainman110 I'll take a look later, thank you for checking it.

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [inline recipes/cminpack/all/conanfile.py:87]:
```suggestion
        self.cpp_info.set_property("cmake_target_name", "cminpack")
```

The idea is transfer the default cmake target to `cminpack`, so, in case you link to the target `cminpack` it will link both libraries. As result, `cminpack::cminpack` should link only to the single library.

I checked locally, and it seems work: 

I enabled `CMAKE_VERBOSE_MAKEFILE` to see the output.

```
----Running------
> cmake --build "/home/conan/project/test_package/build/Release" '--' '-j5'
-----------------
/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -S/home/conan/project/test_package -B/home/conan/project/test_package/build/Release --check-build-system CMakeFiles/Makefile.cmake 0
/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_progress_start /home/conan/project/test_package/build/Release/CMakeFiles /home/conan/project/test_package/build/Release//CMakeFiles/progress.marks
/usr/bin/make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/conan/project/test_package/build/Release'
/usr/bin/make  -f CMakeFiles/cminpack_test_float.dir/build.make CMakeFiles/cminpack_test_float.dir/depend
/usr/bin/make  -f CMakeFiles/cminpack_test_double.dir/build.make CMakeFiles/cminpack_test_double.dir/depend
make[2]: Entering directory '/home/conan/project/test_package/build/Release'
cd /home/conan/project/test_package/build/Release && /opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_depends "Unix Makefiles" /home/conan/project/test_package /home/conan/project/test_package /home/conan/project/test_package/build/Release /home/conan/project/test_package/build/Release /home/conan/project/test_package/build/Release/CMakeFiles/cminpack_test_double.dir/DependInfo.cmake --color=
make[2]: Entering directory '/home/conan/project/test_package/build/Release'
cd /home/conan/project/test_package/build/Release && /opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_depends "Unix Makefiles" /home/conan/project/test_package /home/conan/project/test_package /home/conan/project/test_package/build/Release /home/conan/project/test_package/build/Release /home/conan/project/test_package/build/Release/CMakeFiles/cminpack_test_float.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/conan/project/test_package/build/Release'
make[2]: Leaving directory '/home/conan/project/test_package/build/Release'
/usr/bin/make  -f CMakeFiles/cminpack_test_float.dir/build.make CMakeFiles/cminpack_test_float.dir/build
/usr/bin/make  -f CMakeFiles/cminpack_test_double.dir/build.make CMakeFiles/cminpack_test_double.dir/build
make[2]: Entering directory '/home/conan/project/test_package/build/Release'
make[2]: Entering directory '/home/conan/project/test_package/build/Release'
[ 50%] Linking C executable cminpack_test_double
[ 50%] Linking C executable cminpack_test_float
/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_link_script CMakeFiles/cminpack_test_float.dir/link.txt --verbose=1
/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_link_script CMakeFiles/cminpack_test_double.dir/link.txt --verbose=1
/usr/bin/gcc -m64 -O3 -DNDEBUG -m64 CMakeFiles/cminpack_test_double.dir/tchkderc.c.o -o cminpack_test_double  /home/conan/.conan/data/cminpack/1.3.8/_/_/package/19729b9559f3ae196cad45cb2b97468ccb75dcd1/lib/libcminpack_s.a -lm 
/usr/bin/gcc -m64 -O3 -DNDEBUG -m64 CMakeFiles/cminpack_test_float.dir/tchkderc.c.o -o cminpack_test_float  /home/conan/.conan/data/cminpack/1.3.8/_/_/package/19729b9559f3ae196cad45cb2b97468ccb75dcd1/lib/libcminpacks_s.a -lm 
make[2]: Leaving directory '/home/conan/project/test_package/build/Release'
make[2]: Leaving directory '/home/conan/project/test_package/build/Release'
[100%] Built target cminpack_test_float
[100%] Built target cminpack_test_double
make[1]: Leaving directory '/home/conan/project/test_package/build/Release'
/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_progress_start /home/conan/project/test_package/build/Release/CMakeFiles 0
```

Comment (rainman110) [reply recipes/cminpack/all/conanfile.py:87]: Wow... it really seems to work. But I still don't understand why!

Unfortunately, I could not really follow your explanation. 

> ...  in case you link to the target cminpack it will link both libraries.

??? Do you mean cminpack without namespace prefix? Ideally, this target is deleted, but hey.... we probably cannot have everything 😏 

Awesome job!

**Comment by rainman110** [comment]
Ready for final review

File: recipes/cminpack/all/conanfile.py
Comment (uilianries) [reply recipes/cminpack/all/conanfile.py]: >  Do you mean cminpack without namespace prefix

Yes! But you can use another target name, if you want.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**
